package/endpoint_adapter/src/endpoint_adapter.c

```diff
@@ -532,53 +537,91 @@ static ssize_t fd_read(int fd, void *buffer, size_t count)
   }
 }
 
-#define MSG_ERROR_SERIAL_FLUSH "Interface %s output buffer is full. Dropping data."
+static bool needs_outq_check(int fd)
+{
+  return isatty(fd) && outq > 0;
+}
 
-static ssize_t fd_write(int fd, const void *buffer, size_t count)
+static bool ensure_outq_space(int fd, size_t count)
 {
-  if (isatty(fd) && (outq > 0)) {
-    int qlen;
+  int qlen;
+  ioctl(fd, TIOCOUTQ, &qlen);
+  if (qlen + count > outq) {
+    /* Flush the output buffer, otherwise we'll get behind and start
+     * transmitting partial SBP packets, we must drop some data here, so we
+     * choose to drop old data rather than new data.
+     */
+    tcflush(fd, TCOFLUSH);
     ioctl(fd, TIOCOUTQ, &qlen);
-    if (qlen + count > outq) {
-      /* Flush the output buffer, otherwise we'll get behind and start
-       * transmitting partial SBP packets, we must drop some data here, so we
-       * choose to drop old data rather than new data.
-       */
-      tcflush(fd, TCOFLUSH);
-      ioctl(fd, TIOCOUTQ, &qlen);
-      if (qlen != 0) {
-        if (strstr(port_name, "usb") != port_name) {
-          piksi_log(LOG_WARNING, "Could not completely flush tty: %d bytes remaining.", qlen);
-        } else {
-          /* USB gadget serial can't flush properly for some reason, ignore...
-           *   (This is ignored ad infinitum because this condition occurs on
-           *   start-up before the interface is read from, after the interface
-           *   is read from, it never occurs again.)
-           */
-          return count;
-        }
+    if (qlen != 0) {
+      if (strstr(port_name, "usb") != port_name) {
+        piksi_log(LOG_WARNING, "Could not completely flush tty: %d bytes remaining.", qlen);
+      } else {
+        /* USB gadget serial can't flush properly for some reason, ignore...
+         *   (This is ignored ad infinitum because this condition occurs on
+         *   start-up before the interface is read from, after the interface
+         *   is read from, it never occurs again.)
+         */
+        return false;
       }
-      piksi_log(LOG_ERR, MSG_ERROR_SERIAL_FLUSH, port_name);
-      sbp_log(LOG_ERR, MSG_ERROR_SERIAL_FLUSH, port_name);
-      return count;
     }
+    piksi_log(LOG_ERR | LOG_SBP, "Interface %s output buffer is full. Dropping data.", port_name);
+    return false;
   }
-  while (1) {
-    ssize_t ret = write(fd, buffer, count);
+  return true;
+}
+
+static ssize_t fd_write_with_timeout(int handle,
+                                     const void *buffer,
+                                     size_t count,
+                                     const size_t max_send_sleep_ms,
+                                     const size_t per_retry_sleep_ns)
+{
+  assert(per_retry_sleep_ns > 0);
+  const size_t max_send_sleep_count = MS_TO_NS(max_send_sleep_ms) / per_retry_sleep_ns;
```

Comment:
truncation

Reply:
Truncation?

Reply:
`x.999` will floor to `x`. Does it matter?

Reply:
I don't think it matters for this case